src/runtime/importhook.cs

```diff
@@ -115,187 +97,154 @@ internal static void SaveRuntimeData(RuntimeDataStorage storage)
 
         internal static void RestoreRuntimeData(RuntimeDataStorage storage)
         {
-            InitImport();
             storage.GetValue("py_clr_module", out py_clr_module);
             var rootHandle = storage.GetValue<IntPtr>("root");
             root = (CLRModule)ManagedType.GetManagedObject(rootHandle);
+            BorrowedReference dict = Runtime.PyImport_GetModuleDict();
+            Runtime.PyDict_SetItemString(dict.DangerousGetAddress(), "clr", py_clr_module);
+            SetupNamespaceTracking();
         }
 
-        /// <summary>
-        /// Return the clr python module (new reference)
-        /// </summary>
-        public static unsafe NewReference GetCLRModule(BorrowedReference fromList = default)
+        static void SetupImportHook()
         {
-            root.InitializePreload();
-
-            // update the module dictionary with the contents of the root dictionary
-            root.LoadNames();
-            BorrowedReference py_mod_dict = Runtime.PyModule_GetDict(ClrModuleReference);
-            using (var clr_dict = Runtime.PyObject_GenericGetDict(root.ObjectReference))
+            // Create the import hook module
+            var import_hook_module = Runtime.PyModule_New("clr.loader");
+
+            // Run the python code to create the module's classes.
+            var builtins = Runtime.PyEval_GetBuiltins();
+            var exec = Runtime.PyDict_GetItemString(builtins, "exec");
+            using var args = NewReference.DangerousFromPointer(Runtime.PyTuple_New(2));
+
+            IntPtr codeStr = Runtime.PyString_FromString(LoaderCode);
+            Runtime.PyTuple_SetItem(args.DangerousGetAddress(), 0, codeStr);
+            // PyTuple_SetItem steals a reference, mod_dict is borrowed.
+            var mod_dict = Runtime.PyModule_GetDict(import_hook_module);
+            Runtime.XIncref(mod_dict.DangerousGetAddress());
+            Runtime.PyTuple_SetItem(args.DangerousGetAddress(), 1, mod_dict.DangerousGetAddress());
+            Runtime.PyObject_Call(exec.DangerousGetAddress(), args.DangerousGetAddress(), IntPtr.Zero);
```

Comment:
It is better to make an overload with the new reference types when possible without rewriting existing code, than to call `Dangerous*` methods. See for example https://github.com/pythonnet/pythonnet/blob/7d8f754ff6c4787133427f707f58727d4ddc4981/src/runtime/runtime.cs#L996

One can also be made for `PyTuple_SetItem` using the new `StolenReference`.

Reply:
This applies to a few Runtime methods across the PR. Basically anything where you can add a new overload.